File: C++/Solution38.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution38 {
public:
    string countAndSay(int n) {
        if(n == 1){
            return "1";
        }
        string str = countAndSay(n - 1);
        string result = "";
        int index = 0, number = 0;
        char cur = str[0];
        while(index < str.length()){
            if(str[index] == cur){
                index++;
                number++;
            }
            else{
                result += number + '0';
                result += cur;
                cur = str[index];
                number = 1;
                index++;
            }
        }
        result += number + '0';
        result += cur;
        return result;
    }
};

File: Java/string/Solution38.java
class Solution38 {
    public String countAndSay(int n) {
    	if(n == 1){
    		return "1";
    	}
    	return describe(countAndSay(n - 1));
    }

    private String describe(String s){
    	String result = "";
    	int count = 0;
    	char number = ' ';
    	for(int i = 0; i < s.length(); i++){
    		char cur = s.charAt(i);
    		if(count == 0){
    			number = cur;
    			count = 1;
    		}

    		else if(cur != number){
    			result += String.valueOf(count);
    			result += number;
    			number = cur;
    			count = 1;
    		}

    		else{
    			count += 1;
			}

    		if(i == s.length() - 1){
    			result += String.valueOf(count);
    			result += number;
    		}
    	}
    	return result;
    }

    public static void main(String[] args){
    	Solution38 s = new Solution38();
    	System.out.println(s.describe("111"));
    	System.out.println(s.describe("2112"));
    }
}